### ISSUE TITLE
data augmentation in Keras for large datasets

### ISSUE THREAD

**Issue by prashanthdumpuri** [keras-team/keras#10776]
I'm using Keras to train a model for image classification and 'am working with ~50k images. Each image has three channels and size of each image is 150x150. I have to use floats to store the images because of the minute differences in image intensities between the three channels. I'm using a GPU for training but I do not have a lot of memory on my graphics card and neither do I have the monies to upgrade my GPU. I also have to augment my dataset because my training images do not cover all the possible rotations and translations in my testing dataset. I have written my own generator that splits the input images and labels into chunks before feeding it to Keras' data augmentation routine and model.fit(). Below is my code:

	 from __future__ import print_function
	 from keras.preprocessing.image import ImageDataGenerator
	 from keras.utils import np_utils
	 from keras.callbacks import Callback
	 from keras.callbacks import ModelCheckpoint
	 from keras.callbacks import ReduceLROnPlateau
	 from keras.callbacks import CSVLogger
	 from keras.callbacks import EarlyStopping, TensorBoard, LearningRateScheduler
	 from keras.optimizers import SGD, Adam, RMSprop
	 from keras import backend as K
	 import tensorflow as tf
	 from sklearn.model_selection import train_test_split

	 import numpy as np
	 import math
	 import myCNN # my own convolutional neural network

	def myBatchGenerator(X_train_large, y_train_large, chunk_size):
		number_of_images = len(y_train_large)

		while True:
			batch_start = 0
			batch_end = chunk_size

			while batch_start < number_of_images:
				limit = min(batch_end, number_of_images)
				X = X_train_large[batch_start:limit,:,:,:]
				y = y_train_large[batch_start:limit,:]
				yield(X,y)

				batch_start += chunk_size
				batch_end += chunk_size

	if __name__ == '__main__':
		input_image_shape = (150,150,3)
		# read input images and labels
		# X_train_large is an array of type float16          
		# y_train_large is an array of size number of images x number of classes 
		X_train_large, y_train_large = myFunctionToReadTrainingImagesAndLabels()

		# validation images: about 5000 images 
		X_validation_large, y_validation_large =  myFunctionToReadValidationImagesAndLabels() 
		# create a stratified sample from the large training set. use 100 samples from each class
		y_train_large_vectors = [np.where(r == 1)[0][0] for r in y_train_large]
		unique, counts = np.unique(y_train_large_vectors, return_counts=True)

		X_train_sample = np.empty((12000, 150, 150, 3))
		y_train_sample = np.empty((12000, 12))

		for idx in range(num_classes):
			start_idx_for_sample = 100*idx
			end_idx_for_sample = start_idx_for_sample+99
			start_idx_for_large = np.max(counts)*idx
			end_idx_for_large = start_idx_for_large+99

			X_train_sample[start_idx_for_sample:end_idx_for_sample,:,:,:] = X_train_large[start_idx_for_large:end_idx_for_large,:,:,:]
			y_train_sample[start_idx_for_sample:end_idx_for_sample,:] = y_train_large[start_idx_for_large:end_idx_for_large,:]

		# define augmentation needed for image data generator
		train_datagen = ImageDataGenerator(featurewise_center=False,  
                                             samplewise_center=False,  
					    featurewise_std_normalization=False, 
					    samplewise_std_normalization=False,  
					    zca_whitening=False,  
					    rotation_range=90,  
					    width_shift_range=0.1,  
					    height_shift_range=0.1, 
					    horizontal_flip=True, 
					    vertical_flip=True)  										   
		train_datagen.fit(X_train_sample)
		
		# load my model
		model = myCNN.build_model(input_image_shape)
		sgd = SGD(lr=0.05,decay=10e-4,momentum=0.9)
		model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'
		
		for e in range(number_of_epochs):
		   print('*********************epoch',e)
		   # get 1000 images at a time from the input image set
		   for X_train, y_train in myBatchGenerator(X_train_large, y_train_large,chunk_size=1000):
			   # split it into batches of 32 images/labels and augment on the fly
			   for X_batch, y_batch in train_datagen.flow(X_train_large,y_train_large,batch_size=32):
				   # train
				   model.fit(X_batch,y_batch,validation_data=(X_validation_large,y_validation_large))

		model.save('myCNN_trained_on_largedataset.h5') 

In short, 1. I create a stratified sample of my input images to use for the image data generator. 2. I split my input images into chunks of 1000 images and feed those 1000 images to the model in batches of 32.

So, I'm training my model on 32 images at a time, augmenting it on the fly and 'am validating the model on ~5000 images.

I'm still running my model but each batch of 32 images is currently taking 30 seconds to solve. This translates to a lot of hours to solve just one epoch. I'm missing something here.

I've tested my CNN code on a smaller dataset and it works. So I know the problem is not my function to read input images nor my CNN. I think it is how am splitting my data into chunks and batching it. But I cannot figure out where I went wrong. Can you please guide me?

Thanks in advance for your time
